### PR TITLE
Add return bytes to flashloan callback

### PR DESCRIPTION
**Pull request by ultrasecreth**
Adds return bytes to the flash loan callback and the flash loan function so a callback user can pass data back to itself without using storage.

### CODE REVIEW

File: src/Morpho.sol [hunk 1]
```diff
@@ -399,12 +399,12 @@ contract Morpho is IMorpho {
     /* FLASH LOANS */
 
     /// @inheritdoc IMorpho
-    function flashLoan(address token, uint256 assets, bytes calldata data) external {
+    function flashLoan(address token, uint256 assets, bytes calldata data) external returns (bytes memory resultData) {
```
Comment (MerlinEgalite) [inline src/Morpho.sol:402]:
```suggestion
    function flashLoan(address token, uint256 assets, bytes calldata data) external returns (bytes memory returnData) {
```

File: src/Morpho.sol [hunk 2]
```diff
@@ -399,12 +399,12 @@ contract Morpho is IMorpho {
...
         IERC20(token).safeTransfer(msg.sender, assets);
 
         emit EventsLib.FlashLoan(msg.sender, token, assets);
 
-        IMorphoFlashLoanCallback(msg.sender).onMorphoFlashLoan(assets, data);
+        resultData = IMorphoFlashLoanCallback(msg.sender).onMorphoFlashLoan(assets, data);
```
Comment (MerlinEgalite) [inline src/Morpho.sol:407]:
```suggestion
        returnData = IMorphoFlashLoanCallback(msg.sender).onMorphoFlashLoan(assets, data);
```

File: src/interfaces/IMorpho.sol [hunk 3]
```diff
@@ -280,7 +280,8 @@ interface IMorpho {
     /// @param token The token to flash loan.
     /// @param assets The amount of assets to flash loan.
     /// @param data Arbitrary data to pass to the `onMorphoFlashLoan` callback.
-    function flashLoan(address token, uint256 assets, bytes calldata data) external;
+    /// @return resultData Arbitrary data returned by the callback.
+    function flashLoan(address token, uint256 assets, bytes calldata data) external returns (bytes memory resultData);
```
Comment (MerlinEgalite) [inline src/interfaces/IMorpho.sol:284]:
```suggestion
    /// @return returnData Arbitrary data returned by the callback.
    function flashLoan(address token, uint256 assets, bytes calldata data) external returns (bytes memory returnData);
```

File: src/interfaces/IMorphoCallbacks.sol [hunk 4]
```diff
@@ -48,5 +48,6 @@ interface IMorphoFlashLoanCallback {
     /// @dev The callback is called only if data is not empty.
     /// @param assets The amount of assets that was flash loaned.
     /// @param data Arbitrary data passed to the `flashLoan` function.
-    function onMorphoFlashLoan(uint256 assets, bytes calldata data) external;
+    /// @return resultData Arbitrary data returned by the callback.
+    function onMorphoFlashLoan(uint256 assets, bytes calldata data) external returns (bytes memory resultData);
```
Comment (MerlinEgalite) [inline src/interfaces/IMorphoCallbacks.sol:52]:
```suggestion
    /// @return returnData Arbitrary data returned by the callback.
    function onMorphoFlashLoan(uint256 assets, bytes calldata data) external returns (bytes memory returnData);
```

File: src/mocks/FlashBorrowerMock.sol [hunk 5]
```diff
@@ -12,13 +12,14 @@ contract FlashBorrowerMock is IMorphoFlashLoanCallback {
         MORPHO = newMorpho;
     }
 
-    function flashLoan(address token, uint256 assets, bytes calldata data) external {
-        MORPHO.flashLoan(token, assets, data);
+    function flashLoan(address token, uint256 assets, bytes calldata data) external returns (bytes memory resultData) {
```
Comment (MerlinEgalite) [inline src/mocks/FlashBorrowerMock.sol:15]:
```suggestion
    function flashLoan(address token, uint256 assets, bytes calldata data) external returns (bytes memory returnData) {
```

File: src/mocks/FlashBorrowerMock.sol [hunk 6]
```diff
@@ -12,13 +12,14 @@ contract FlashBorrowerMock is IMorphoFlashLoanCallback {
...
+        return MORPHO.flashLoan(token, assets, data);
     }
 
-    function onMorphoFlashLoan(uint256 assets, bytes calldata data) external {
+    function onMorphoFlashLoan(uint256 assets, bytes calldata data) external returns (bytes memory resultData) {
```
Comment (MerlinEgalite) [inline src/mocks/FlashBorrowerMock.sol:19]:
```suggestion
    function onMorphoFlashLoan(uint256 assets, bytes calldata data) external returns (bytes memory returnData) {
```

File: test/forge/integration/CallbacksIntegrationTest.sol [hunk 7]
```diff
@@ -60,13 +60,14 @@ contract CallbacksIntegrationTest is
         }
     }
 
-    function onMorphoFlashLoan(uint256 amount, bytes memory data) external {
+    function onMorphoFlashLoan(uint256 amount, bytes memory data) external returns (bytes memory resultData) {
```
Comment (MerlinEgalite) [inline test/forge/integration/CallbacksIntegrationTest.sol:63]:
```suggestion
    function onMorphoFlashLoan(uint256 amount, bytes memory data) external returns (bytes memory returnData) {
```

File: test/forge/integration/CallbacksIntegrationTest.sol [hunk 8]
```diff
@@ -60,13 +60,14 @@ contract CallbacksIntegrationTest is
...
         require(msg.sender == address(morpho));
         bytes4 selector;
         (selector, data) = abi.decode(data, (bytes4, bytes));
         if (selector == this.testFlashLoan.selector) {
             assertEq(loanToken.balanceOf(address(this)), amount);
             loanToken.approve(address(morpho), amount);
+            resultData = "Arbitrary data!";
```
Comment (MerlinEgalite) [inline test/forge/integration/CallbacksIntegrationTest.sol:70]:
```suggestion
            returnData = "Arbitrary data!";
```

**Comment by Rubilmax** [comment]
Former discussions for reference:
- https://github.com/morpho-org/morpho-blue/pull/190
- https://github.com/morpho-org/morpho-blue/pull/168#issuecomment-1649545002
- https://github.com/morpho-org/morpho-blue/issues/16#issuecomment-1632161962

File: test/forge/integration/CallbacksIntegrationTest.sol [hunk 9]
```diff
@@ -78,9 +79,11 @@ contract CallbacksIntegrationTest is
         loanToken.setBalance(address(this), amount);
         morpho.supply(marketParams, amount, 0, address(this), hex"");
 
-        morpho.flashLoan(address(loanToken), amount, abi.encode(this.testFlashLoan.selector, hex""));
+        bytes memory resultData =
```
Comment (MerlinEgalite) [inline test/forge/integration/CallbacksIntegrationTest.sol:82]:
```suggestion
        bytes memory returnData =
```

File: test/forge/integration/CallbacksIntegrationTest.sol [hunk 10]
```diff
@@ -78,9 +79,11 @@ contract CallbacksIntegrationTest is
...
+            morpho.flashLoan(address(loanToken), amount, abi.encode(this.testFlashLoan.selector, hex""));
 
         assertEq(loanToken.balanceOf(address(morpho)), amount, "balanceOf");
+        assertEq(resultData, "Arbitrary data!", "resultData");
```
Comment (MerlinEgalite) [inline test/forge/integration/CallbacksIntegrationTest.sol:86]:
```suggestion
        assertEq(returnData, "Arbitrary data!", "returnData");
```

**Review by MerlinEgalite — COMMENTED** [review]
Alright I'm not against this change indeed but then should we harmonize it for other callbacks? 

I'm against for the other callbacks.

**Pull request merged**